### ISSUE TITLE
appVersion tag

### ISSUE THREAD

**Issue by kfox1111** [helm/helm#7517]
Newer versions of 'helm create' spit out a template that includes:
```
image: "{{ .Values.image.repository }}:{{ .Chart.AppVersion }}"
```

This is nice for the release process, but provides no way for the end user to override the tag when necessary (sometimes for mirroring).

There should be a way to set image.tag to something or else fall back to appVersion.

**Comment by mattfarina** [comment]
It would be nice if there was a good pattern to override the image. Either with a pattern in the `values.yaml` file or a flag like the one described on #3555.

**Comment by maver1ck** [comment]
Flag is not the best solution as it's difficult to versioning.

I prefer values.yaml.

Maybe just:
```
Chart:
  AppVersion: x.y.z
```

**Comment by kfox1111** [comment]
Flag has an advantage when used with package.
helm package --app-version 1.2.0 chart/
Could easily be used in a pipeline to autogenerate packages writing in the appVersion into the chart.yaml and put the new package in a registry.

**Comment by TBBle** [comment]
There's a few different common patterns in values.yaml. For example Skaffold supports three of them in its [`imageStrategy`](https://skaffold.dev/docs/references/yaml/) configuration option:
* Single value override
* Split value overrides for `repository` and `tag`, e..g. [cert-manager](https://github.com/jetstack/cert-manager/blob/930827b06bc03afd021fde49907095eac4784d0d/deploy/charts/cert-manager/values.yaml#L37), and [jaeger-operator](https://github.com/jaegertracing/helm-charts/blob/08092b312723dd5c52c80a24a9ada4fb66343bd4/charts/jaeger-operator/values.yaml#L4)
  * [Agones](https://github.com/googleforgames/agones/blob/619867f7d6e495da982c807e34e9c8c2439d320c/install/helm/agones/values.yaml#L126) uses `registry` and `tag` for the same split, which won't work with Skaffold.
* Split value overrides for `registry`, `repository`, and `tag`, e.g., [postgresql](https://github.com/helm/charts/blob/0b64349aeb537d0fd038df0e1d40ec5cf1206609/stable/postgresql/values.yaml#L15). I suspect all the charts in https://github.com/helm/charts have standardised on this now. We use this format at my work too, because we can feed `registry` from a global value very easily, to support mirrors of our Docker images on less-secure networks.

The current default starter only supports an override for `repository` but not `tag`, so it wouldn't work with Skaffold, for example, as Skaffold works by overriding the tag to point to the SHA256 of its generated image.

At my work, (roughly, I'm going by memory) we have our Deployment `image` be something like
```
image: "{{ .Values.image.registry }}/{{ .Values.image.repository }}:{{ tpl .Values.image.tag . }}"
```

and our values.yaml defaults `image.tag` to `{{ .Chart.Version }}`. Our CI script changes `Chart.Version`, `Chart.AppVersion` and `image.tag` in values.yaml before test-install and before packaging, so we get autobuilt charts using Git-SHA-tagged containers (because Docker doesn't allow some SemVer characters such as `+` in tags...), and then for release builds we just don't override those values, and it picks up our release-tagged containers.

[cert-manager](https://github.com/jetstack/cert-manager/blob/930827b06bc03afd021fde49907095eac4784d0d/deploy/charts/cert-manager/templates/deployment.yaml#L75) achieves a similar effect using `default` rather than `tpl`, which is probably a better approach, on reflection.

This might have been made somewhat simpler through something like #2492, but per https://github.com/helm/helm/issues/2492#issuecomment-413635332 this wasn't done in favour of the (not yet done?) Lua-based event hooks in Helm v3. And we need to be installable with Helm v2 for now anyway.

**Comment by kfox1111** [comment]
The problem with values, is they wouldn't be captured during a package. It would be harder to override the values.yaml file during package then poking the updated value in Chart.yaml?

**Comment by TBBle** [comment]
I would suggest supporting both. It's been very handy for us to be able to override an image tag at `helm install` or `helm upgrade` time. For example, when pushing a hotfix to a customer that didn't change the helm chart, they can deploy just a new image with `--set subchart.image.tag=hotfix-tag`.

What I described above (and what _cert-manager_ does) supports both values override and `helm package` flag override, as if you don't set the value, you get `.Chart.Version` or `Chart.AppVersion` by default. Then you can use `--version` or `--appversion` with `helm package` to effectively override the tag at runtime.

This doesn't work for the use case I have here right now, because Docker tags cannot hold an entire semver (no `+`), and because we install the package from source tree as part of our testing pipeline, so we need to override the `version` and `appVersion` in Charts.yaml for that reason anyway.

That said, I've gathered a few ideas from here that I might consider applying. If I stuck the git SHA into the appVersion of autobuilds, and deployed test from a `helm package` tgz instead of from source, a few other pieces of the pipeline become much simpler as well, and I can drop some `sed` calls from our CI script.

@TBBle: Follow up the week.

**Comment by timothyclarke** [comment]
> Flag has an advantage when used with package.
> helm package --app-version 1.2.0 chart/
> Could easily be used in a pipeline to autogenerate packages writing in the appVersion into the chart.yaml and put the new package in a registry.

@kfox1111 It's the other end of the process (install / upgrade) that is causing most people the pain. When you're packaging a chart you can easily modify the appVersion. 
eg 
```
sed -i "/appVersion:/c\appVersion: ${SEM_VER}" chart/Chart.yaml
helm package chart/
```

In my experience it's when people want to install / upgrade that they want to use a specific docker tag.
Note the code segment provided is part of `templates/deployment.yaml`

**Comment by kfox1111** [comment]
Both use cases need to be covered I think..

I have had need to override appVersion at install/upgrade time for sure.

That sed line works, and I've used it, but is kind of a hack. the same hack could be used for installing/upgrading as well, but its also ugly for the user:
```
helm pull foo/chart --untar
sed -i "/appVersion:/c\appVersion: ${SEM_VER}" chart/Chart.yaml
helm install chart ....
```
I'm not recommending it.

If charts are going to use appVersion in their templates, upgrade/install/package all should have a simple way to override it.

**Comment by timothyclarke** [comment]
> same hack could be used for installing/upgrading as wel
Only if you download and edit the chart. If you're going to mandate that people need to download and edit the chart which begs the question why does helm allow you to install a chart from a remote repo without forcing the download and unpack first ?

The other approach is why use `appVersion` at all? The older methods used `image.tag` which could be used at deployment time.

I completely agree with your statement about overriding everywhere.  I added more detail in the #3652 thread referenced above

**Comment by TBBle** [comment]
After applying the discussions in this thread, we ended up with
```
{{- $tag := .Values.image.tag | default .Chart.AppVersion | default .Chart.Version -}}
```

I'd suggest the default template do that too.

It turns out `.Chart.AppVersion` is optional, and we don't use it for releases, as it equals the `.Chart.Version`, so that's one fewer places to update when bumping versions after a release. For auto-builds, we set `.Chart.AppVersion` to the git commit sha (using `helm package --app-version`), which matches our autobuild Docker tag, and this leaves `.Values.image.tag` as available for install-time override if desired.

**Comment by forgo** [comment]
@TBBle Just FYI, our team recently upgraded to Helm 3.

In an attempt to create a new Helm 3 chart for a new application, I ran into this issue while using the `jib` gradle deploy mechanism in Skaffold.

Skaffold is supposed to autogenerate an image with a custom local tag with a suffiix like `-dirty-12345` and override the helm values with that.

Anybody using skaffold to pick up local image builds like this is going to run into this issue when creatiing new helm 3 charts.

 If the Skaffold behavior should change in the future, you might want to coordinate with them; otherwise, for now, your suggested fix worked for put putting it inline in my auto-generated `deployment.yaml`:

```
spec:
  containers:
    - name: {{ .Chart.Name }}
      securityContext:
        {{- toYaml .Values.securityContext | nindent 12 }}
      image: "{{ .Values.image.repository }}:{{ .Values.image.tag | default .Chart.AppVersion | default .Chart.Version }}"
      imagePullPolicy: {{ .Values.image.pullPolicy }}
```

**Comment by timothyclarke** [comment]
I'd imagine that conifg snippet would best sit in `_helpers.tpl`

**Comment by TBBle** [comment]
That's where our local chart has the example I pasted above. I haven't created a Helm chart from scratch for a while, so I wasn't sure if there _was_ a _helpers.tpl in the default starter.

**Comment by bacongobbler** [comment]
closed with #7878.

**Comment by kfox1111** [comment]
What about the helm package use case?

**Comment by bacongobbler** [comment]
That is covered in #3555.

**Comment by bacongobbler** [comment]
> What about the helm package use case?

Additionally, `helm package --app-version` already exists and is available in Helm 3.1.2.

**Comment by kfox1111** [comment]
Doh. I missed that somehow. Sorry. Thanks for helping me see it.
